tests/test_mutations.py

```diff
@@ -472,47 +705,288 @@ def test_keep_mutation_parent_zero_rate(self):
+                mut = mutations[j]
+                mutation_item = MutationItem(
+                        site=mut.site,
+                        node=mut.node,
+                        derived_state=mut.derived_state,
+                        parent=mut.parent,
+                        metadata=mut.metadata,
+                        time=nodes.time[mut.node],
+                        new=False,
+                        id=tskit.NULL)
+                mutation_list.append(mutation_item)
+                j += 1
+
+            site_item = SiteItem(
+                    position=site.position,
+                    ancestral_state=site.ancestral_state,
+                    metadata=site.metadata,
+                    mutations=mutation_list)
+            self.sites.append(site_item)
+
+    def populate_tables(self, tables):
+        for site in self.sites:
+            site_id = tables.sites.add_row(
+                    site.position,
+                    site.ancestral_state,
+                    site.metadata)
+            for mutation in site.mutations:
+                tables.mutations.add_row(
+                            site_id, mutation.node, mutation.derived_state,
+                            parent=mutation.parent, metadata=mutation.metadata)
+
+    def place_mutations(self, tables, rng, discrete_sites=False):
+        # Insert a sentinel into the map for convenience.
+        map_position = np.hstack([self.rate_map.position, [tables.sequence_length]])
+        node_times = tables.nodes.time
+        for edge in tables.edges:
+            branch_start = node_times[edge.child]
+            branch_end = node_times[edge.parent]
+            branch_length = branch_end - branch_start
+            index = np.searchsorted(map_position, edge.left)
+            if map_position[index] > edge.left:
+                index -= 1
+            left = edge.left
+            right = 0
+            while right != edge.right:
+                right = min(edge.right, map_position[index + 1])
+                site_left = np.ceil(left) if discrete_sites else left
+                site_right = np.ceil(right) if discrete_sites else right
+                assert site_left <= site_right
+                assert map_position[index] <= left
+                assert right <= map_position[index + 1]
+                assert right <= edge.right
+                # Generate the mutations.
+                rate = self.rate_map.rate[index]
+                mu = rate * (site_right - site_left) * branch_length
+                for _ in range(rng.poisson(mu)):
+                    position = rng.flat(site_left, site_right)
+                    if discrete_sites:
+                        position = np.floor(position)
+                    assert edge.left <= position
+                    assert position < edge.right
+                    time = rng.flat(branch_start, branch_end)
+                    site_positions = [s.position for s in self.sites]
```

Comment:
We can get rid of this O(n) with
```python
if position in self.sites:
    site = self.sites[position]
else:
     # new site

Reply:
Well, this won't happen in the C code...